ucp/useraccounts_module.php

```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
-				{
-					$sql = "SELECT b.badge_id, b.badge_label, b.badge_icon_24, bt.badge_type_name
-						FROM $badge_table AS b, $badge_type_table AS bt
-						WHERE b.badge_default = TRUE
-						AND b.badge_type_id = bt.badge_type_id
-						UNION
-						SELECT b.badge_id, b.badge_label, b.badge_icon_24, bt.badge_type_name
-						FROM $badge_table AS b, $badge_type_table AS bt, $badge_avail_table AS ba
-						WHERE (b.badge_id = ba.badge_id AND ba.user_id = $this_user_id)
-						AND b.badge_type_id = bt.badge_type_id
-						ORDER BY badge_type_name";
-					if (!($result = $this->db->sql_query($sql)))
-					{
-						trigger_error($this->user->lang('UCP_AVAIL_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
-					}
-					// Prepare an array to group badges by types.
-					$typelist = array();
-					// Loop over available badges and highlight those that are selected by the user.
-					while ($row = $this->db->sql_fetchrow($result))
+					// Check if the type name is in the typelist, if not, add it to the list and to the blockvars for the current row.
+					if (!in_array($row['badge_type_name'], $typelist))
 					{
-						$s_badgerow = array(
-							'S_SELECTED' => $this->core->in_2d_array($row['badge_id'], $selected_badges),
-							'BADGE_ICON_URL' => $row['badge_icon_24'],
-							'BADGE_LABEL' => $row['badge_label'],
-							'ID' => $row['badge_id'],
-						);
-						// Check if the type name is in the typelist, if not, add it to the list and to the blockvars for the current row.
-						if (!in_array($row['badge_type_name'], $typelist))
-						{
-							$typelist[] = $row['badge_type_name'];
-							$s_badgerow['BADGE_TYPE'] = $row['badge_type_name'];
-						}
-						$this->template->assign_block_vars('s_badgerow', $s_badgerow);
+						$typelist[] = $row['badge_type_name'];
+						$s_badgerow['BADGE_TYPE'] = $row['badge_type_name'];
 					}
-					$this->db->sql_freeresult($result);
+
+					$this->template->assign_block_vars('s_badgerow', $s_badgerow);
 				}
+				$this->db->sql_freeresult($result);
 
-				// Assign template vars for mode
+				// Assign template vars for mode.
 				$this->template->assign_vars(array(
 					'S_UCP_MODE_SELECT_BADGES' => true,
-					'S_AVAIL_BADGES' => ($total_avail_count > 0)
+					'S_AVAIL_BADGES' => sizeof($typelist) > 0
 				));
 
 				break;
 			}
 
 			case 'order_badges':
 			{
-				// Set up vars for mode
-				$action = $this->request->variable('action', ''); // Reordering does not work without requesting the action var here explictly as string.
-				$item_id = $this->request->variable('id', 0); // 'item_id' of selected element when reordering.
-				$moveup = $action == 'move_up' ? true : false;
-				$movedown = $action == 'move_down' ? true : false;
+				// Set up vars for mode.
+				$action = $this->request->variable('action', ''); // Required to be defined here again, probably because parameters are given through the URL.
```

Comment:
This comment gives me the heebie jeebies.  Can you please find out what is going on here, and fix it if needed?  What is the value of `$action` before and after this line?

Reply:
When using `$request->variable()`, it automatically sets a type for the variable. The code used in the else clause of 

```php
		if (is_array($action))
		{
			list($action, ) = each($action);
		}
		else
		{
			$action = $request->variable('action', '');
		}
```
is bogus and the array check IMO too so I removed both.

This makes it necessary to explicitly request the `$action` var as an array or string, like for the URL parameter:

```php
		U_MOVE_UP' => $this->u_action . '&amp;action=move_up&amp;id=' . $row['item_id']
```